docs/dev/datastream_execution_mode.md

```diff
@@ -0,0 +1,241 @@
+DataStreamSource<String> source = env.fromElements(...);
+
+source.name("source")
+	.map(...).name("map1")
+	.map(...).name("map2")
+	.rebalance()
+	.map(...).name("map3")
+	.map(...).name("map4")
+	.keyBy((value) -> value)
+	.map(...).name("map5")
+	.map(...).name("map6")
+	.sinkTo(...).name("sink");
+```
+
+Operations that imply a 1-to-1 connection pattern between operations, such as
+`map()`, `flatMap()`, or `filter()` can just forward data straight to the next
+operation, which allows these operations to be chained together. This means
+that Flink would not normally insert a network shuffle between them.
+
+Operation such as `keyBy()` or `rebalance()` on the other hand require data to
+be shuffled between different parallel instances of tasks. This induces a
+network shuffle.
+
+For the above example Flink would group operations together as tasks like this:
+
+- Task1: `source`, `map1`, and `map2`
+- Task2: `map3`, `map4`
+- Task3: `map5`, `map6`, and `sink`
+
+And we have a network shuffle between Tasks 1 and 2, and also Tasks 2 and 3.
+
+#### STREAMING Execution Mode
+
+In STREAMING execution mode, all tasks need to be online/running all the time.
+This allows Flink to immediately process new records through the whole
+pipeline, which we need for low-latency stream processing. This also means that
+the TaskManagers that are allotted to a job need to have enough resources to
+run all the tasks at the same time.
+
+Network shuffles are _pipelined_, meaning that records are immediately sent to
+downstream tasks, with some buffering on the network layer. Again, this enables
+low-latency processing.
+
+#### BATCH Execution Mode
+
+In BATCH execution mode, the tasks of a job can be separated into stages that
+can be executed one after another. We can do this because the input is bounded
+and Flink can therefore fully process one stage of the pipeline before moving
+on to the next. In the above example the job would have three stages that
+correspond to the three tasks that are separated by the shuffle barriers.
+
+Instead of sending records immediately to downstream tasks, as explained above
+for STREAMING mode, processing in stages requires Flink to materialize
+intermediate results of tasks to some non-ephemeral storage which allows
+downstream tasks to read them after upstream tasks have already gone off line.
+This will increase the latency of processing but comes with other interesting
+properties. For one, this allows Flink to backtrack to the latest available
+results when a failure happens instead of restarting the whole job. Another
+side effect is that BATCH jobs can execute on fewer resources (in terms of
+available slots at TaskManagers) because the system can execute tasks
+sequentially one after the other.
+
+TaskManagers will keep intermediate results at least as long as downstream
+tasks have not consumed them. After that, they will be kept for as long as
+space allows in order to allow the aforementioned backtracking to earlier
```

Comment:
I think technically we keep them until all consuming pipelined regions have produced their outputs.

Reply:
I added an explanation in parentheses but right now it will point to the non-existing documentation about pipelined regions. That's another TODO. 😅